### PR TITLE
Document field sensitivity class

### PR DESCRIPTION
**Pull request by romainbrenguier**
This gives an overview of how expressions are transformed by functions
from this class.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [na] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/field_sensitivity.h [hunk 1]
```diff
@@ -15,7 +15,36 @@ class namespacet;
 class goto_symex_statet;
 class symex_targett;
 
-/// \brief Control granularity of object accesses
+/// Control granularity of object accesses
+///
+/// [Field sensitivity](\ref field_sensitivityt) is a transformation of the
+/// instructions goto-program which mainly replaces some expressions by symbols
```
Comment (smowton) [inline src/goto-symex/field_sensitivity.h:21]: The interface of `field_sensitivityt` doesn't mention instructions, it transforms expressions.

File: src/goto-symex/field_sensitivity.h [hunk 2]
```diff
@@ -15,7 +15,36 @@ class namespacet;
...
+/// but can also add assignments to the target equations produced by symbolic
+/// execution.
+/// The main goal of this transformation is too allow more constants to be
```
Comment (smowton) [inline src/goto-symex/field_sensitivity.h]: too -> to

File: src/goto-symex/field_sensitivity.h [hunk 3]
```diff
@@ -15,7 +15,36 @@ class namespacet;
...
+/// propagated during symbolic execution.
+/// Note that field sensitivity is not applied as a single pass over the
+/// whole goto program but instead applied as the symbolic execution unfold.
```
Comment (smowton) [inline src/goto-symex/field_sensitivity.h]: unfolds

File: src/goto-symex/field_sensitivity.h [hunk 4]
```diff
@@ -15,7 +15,36 @@ class namespacet;
 class goto_symex_statet;
 class symex_targett;
 
-/// \brief Control granularity of object accesses
+/// Control granularity of object accesses
+///
+/// [Field sensitivity](\ref field_sensitivityt) is a transformation of the
+/// instructions goto-program which mainly replaces some expressions by symbols
+/// but can also add assignments to the target equations produced by symbolic
+/// execution.
```
Comment (smowton) [inline src/goto-symex/field_sensitivity.h:23]: This is quite a vague description, suggest mentioning that it replaces member operators with atomic symbols representing a field when possible, and when it cannot it adds extra assignments to convert whole-struct operations into several writes to individual fields.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 394f0f3d66686ec7c65e6190d8db478df1e78fcf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115085951

File: src/goto-symex/field_sensitivity.h (hunk 4 above)
Comment (tautschnig) [reply src/goto-symex/field_sensitivity.h:23]: +1 - I think we should separate the high-level aim (during member/index expressions into atomic symbols) from implementation details (fixing up in the failing case). That kind of "separation" might just be done by adding a paragraph in between.

**Review by tautschnig — APPROVED** [review]
Thank you for adding documentation!!!

File: src/goto-symex/field_sensitivity.h (hunk 1 above)
Comment (romainbrenguier) [reply src/goto-symex/field_sensitivity.h:21]: Yes but field_assignments mentions assignments and assignments are not expressions. Should I say expressions and assignments?

File: src/goto-symex/field_sensitivity.h (hunk 4 above)
Comment (romainbrenguier) [reply src/goto-symex/field_sensitivity.h:23]: @tautschnig @smowton I added a paragraph with a high level description before getting into the details, can you check it? (last commit)

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 505442da2d4a66918baa37560d5c595ef56b21e8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115553269

**Review by forejtv — APPROVED** [review]

**Pull request merged**